### PR TITLE
support infinite scroll

### LINKED ISSUE #279: Support infinite scrolling
**Is your feature request related to a problem? Please describe.**
Often times large data sets are paged via API instead of returned all at once. We need a UI trigger to fetch the next batch of data.

**Describe the solution you'd like**
Implement the infinite scrolling API for ensemble

**Describe alternatives you've considered**

**Additional context**
Example implementation in flutter https://github.com/EnsembleUI/ensemble/blob/main/lib/layout/list_view.dart

### PR DESCRIPTION
**Pull request by sagardspeed2**
## Describe your changes
Added feature for support of infinite scroll.

## Issue ticket number and link

Closes #
closes #279 

## Checklist before requesting a review

- [x] I have performed a self-review of my code
- [ ] I have added tests
- [x] I have added a changeset `pnpm changeset add`
- [x] I have added example usage in the kitchen sink app

### CODE REVIEW

File: packages/runtime/src/widgets/GridView.tsx [hunk 1]
```diff
@@ -24,14 +30,25 @@ export type GridViewProps = {
...
 
 export const GridView: React.FC<GridViewProps> = ({
   "item-template": { data, name, template },
   styles,
+  ...rest
 }) => {
   const defaultColumnCount = 4;
+  const { resolvedWidgetId } = useWidgetId();
   const { namedData } = useTemplateData({ data, name });
+  const onScrollEndAction = useEnsembleAction(rest.onScrollEnd);
```
Comment (evshi) [inline packages/runtime/src/widgets/GridView.tsx:44]: Destructure from props

File: packages/runtime/src/widgets/GridView.tsx [hunk 2]
```diff
@@ -92,11 +109,42 @@ export const GridView: React.FC<GridViewProps> = ({
...
+    },
+    [onScrollEndActionCallback],
+  );
+
+  useEffect(() => {
+    if (rest.onScrollEnd) {
+      // assign scroll event listener to element
+      document
+        .getElementById(resolvedWidgetId)
+        ?.addEventListener("scroll", handleScrollEvent);
```
Comment (evshi) [inline packages/runtime/src/widgets/GridView.tsx:131]: Use ref here instead of querying document

**Review by evshi — COMMENTED** [review]
We also need the same capability on grids.

**Comment by sagardspeed2** [comment]
@evshi You mean add this functionality on data grid ?

**Comment by evshi** [comment]
Yes

**Comment by sagardspeed2** [comment]
@evshi in order to do that, can you approve this request and also #366, because in data grid, first we need to add feature for scroll and then we can integrate this functionality on it. ?

File: apps/kitchen-sink/src/ensemble/screens/layouts.yaml [hunk 3]
```diff
@@ -24,16 +24,34 @@ View:
...
+                  styles:
+                    maxHeight: 200px
+                    overflow: auto
+                  onScrollEnd:
+                    # here you can add your page update logic
+                    invokeApi:
+                      name: getDummyProducts
+                      onResponse:
+                        executeCode: |
+                          const prevItem = ensemble.storage.get("products")
```
Comment (evshi) [inline apps/kitchen-sink/src/ensemble/screens/layouts.yaml:40]: Make another API call with the next page

File: packages/runtime/src/widgets/GridView.tsx [hunk 4]
```diff
@@ -24,14 +30,27 @@ export type GridViewProps = {
...
 } & EnsembleWidgetProps<GridViewStyles>;
 
 export const GridView: React.FC<GridViewProps> = ({
   "item-template": { data, name, template },
   styles,
+  onScrollEnd,
+  ...rest
 }) => {
   const defaultColumnCount = 4;
+  const { resolvedWidgetId } = useWidgetId();
```
Comment (evshi) [inline packages/runtime/src/widgets/GridView.tsx:43]: Replace with useRegisterBindings and pass in style

File: packages/runtime/src/widgets/GridView.tsx [hunk 5]
```diff
@@ -92,11 +111,43 @@ export const GridView: React.FC<GridViewProps> = ({
...
+    };
+  }, [resolvedWidgetId, rest.onScrollEnd, handleScrollEvent]);
+
   return (
     <div
-      style={{
-        ...(styles?.visible === false ? { display: "none" } : undefined),
-      }}
+      ref={containerRef}
+      style={{ maxHeight: styles?.maxHeight, overflow: styles?.overflow }}
```
Comment (evshi) [inline packages/runtime/src/widgets/GridView.tsx:150]: Apply resolved styles with spread operator

File: packages/runtime/src/widgets/GridView.tsx [hunk 6]
```diff
@@ -92,10 +114,44 @@ export const GridView: React.FC<GridViewProps> = ({
...
+        }
+      }
+    };
+  }, [resolvedWidgetId, onScrollEnd, handleScrollEvent]);
+
   return (
     <div
+      ref={containerRef}
       style={{
-        ...(styles?.visible === false ? { display: "none" } : undefined),
```
Comment (evshi) [inline packages/runtime/src/widgets/GridView.tsx:98]: This line should remain.

File: packages/runtime/src/widgets/GridView.tsx [hunk 7]
```diff
@@ -24,14 +30,30 @@ export type GridViewProps = {
     name: string;
     template: EnsembleWidget;
   };
+  onScrollEnd?: EnsembleAction;
 } & EnsembleWidgetProps<GridViewStyles>;
 
 export const GridView: React.FC<GridViewProps> = ({
   "item-template": { data, name, template },
   styles,
```
Comment (evshi) [inline packages/runtime/src/widgets/GridView.tsx:38]: Don't need to deconstruct styles now it can be passed along with rest

**Review by evshi — COMMENTED** [review]
Looks good, let's apply to grid as well once #366 is merged

**Review by evshi — APPROVED** [review]

**Pull request merged**